### PR TITLE
WIP: Fix of Issue 138 (Change "Node" to "Signature")

### PR DESCRIPTION
**Pull request by dongbohu**
This is a PR in progress which will fix issue #138.

So far it passes all unit tests but the heatmap isn't shown up correctly. I got the following error when loading the heatmap:
```
TypeError: Cannot read property '_id' of undefined
```
You can reproduce the error by going to:
http://165.123.67.202/#/analyze
Search and add some samples, then click the test tube icon on top right corner, select the machine learning model of "Emsemble Adage 300". Then you should be able to see the error message on the console. So I must have made some mistake when modifying the related modules.

All the other modules seem to be working fine.

@mhuyck and @rzelayafavila: Please play with the web UI and tell me anything that is not working. Thanks.

### CODE REVIEW

**Comment by mhuyck** [comment]
Looks like an issue inside Vega. I'm looking now.

**Comment by dongbohu** [comment]
Don't worry about the issues reported by codeclimate for now. I will fix them when this PR is done.

**Comment by mhuyck** [comment]
Here is the problem: the Activity API returns JSON data that still refer to them as `node`s. The data from this API are loaded without modification into the dataset that Vega uses to draw.

> http://165.123.67.202/api/v0/activity/?mlmodel=4&sample=118&format=json

**Comment by mhuyck** [comment]
Looking for a simple way to fix. I think we can revert a few of the changes in the   `$scope.heatmapSpec` and it will work again.

**Comment by dongbohu** [comment]
Ah, I didn't know that. So I must have modified some places that shouldn't be.

File: interface/src/app/analyze/analysis/analysis.js [hunk 1]
```diff
@@ -91,36 +91,36 @@ function AnalysisCtrl($scope, $log, $q, $state, SampleBin) {
...
-        // compute minimum normalized value for each node (across samples)
-        'name': 'node_summary',
+        // compute minimum normalized value for each signature (across samples)
+        'name': 'signature_summary',
         'source': 'activity',
         'transform': [
           {
             'type': 'aggregate',
-            'groupby': ['node'],
+            'groupby': ['signature'],
```
Comment (mhuyck) [inline interface/src/app/analyze/analysis/analysis.js]: change this back to `node`

File: interface/src/app/analyze/analysis/analysis.js [hunk 2]
```diff
@@ -91,36 +91,36 @@ function AnalysisCtrl($scope, $log, $q, $state, SampleBin) {
...
         'transform': [
           {
             'type': 'lookup',
-            'on': 'node_summary',
-            'onKey': 'node',
-            'keys': ['node'],
-            'as': ['node_summary']
+            'on': 'signature_summary',
+            'onKey': 'signature',
+            'keys': ['signature'],
```
Comment (mhuyck) [inline interface/src/app/analyze/analysis/analysis.js]: revert lines 117 and 118 to say `node`

File: interface/src/app/analyze/analysis/analysis.js [hunk 3]
```diff
@@ -129,22 +129,22 @@ function AnalysisCtrl($scope, $log, $q, $state, SampleBin) {
             'as': ['sample_order']
           }, {
             'type': 'lookup',
-            'on': 'nodeOrder',
+            'on': 'signatureOrder',
             'onKey': 'data',
-            'keys': ['node'],
-            'as': ['node_order']
+            'keys': ['signature'],
```
Comment (mhuyck) [inline interface/src/app/analyze/analysis/analysis.js]: revert this line to `node`

File: interface/src/app/analyze/analysis/analysis.js [hunk 4]
```diff
@@ -202,7 +202,7 @@ function AnalysisCtrl($scope, $log, $q, $state, SampleBin) {
         'properties': {
           'enter': {
             // TODO dynamic layout: compute these constants also (see above)
-            'text': {'value': 'node'},
+            'text': {'value': 'signature'},
```
Comment (mhuyck) [inline interface/src/app/analyze/analysis/analysis.js]: this needs to revert to `node` also

Comment (dongbohu) [inline interface/src/app/analyze/analysis/analysis.js]: I made these changes and I can see the heatmap now! But when I click "Cluster Signatures" button, I got this error:
```
TypeError: Cannot read property 'value' of undefined
```
I suspect this is due to some similar replacement. Can you take a look?

Comment (mhuyck) [inline interface/src/app/analyze/analysis/analysis.js]: Okay, I'm on it.

File: interface/src/app/analyze/analysis/sampleBin.js [hunk 5]
```diff
@@ -144,39 +144,39 @@ MathFuncts, errGen) {
...
-      //      by walking through the `firstSampleNodes` and constructing a
-      //      `nodeObject` for each. [outer .map()]
-      var retval = firstSampleNodes.map(function(val) {
-        var nodeObject = {
+      // (2a) next, we build a new array (`retval`) comprised of
+      //      `signatureObject`s by walking through the `firstSampleSignatures`
+      //       and constructing a signatureObject` for each. [outer .map()]
+      var retval = firstSampleSignatures.map(function(val) {
+        var signatureObject = {
           'id': val.node,
```
Comment (dongbohu) [inline interface/src/app/analyze/analysis/sampleBin.js]: I think we should keep `node` on this line, right?

File: interface/src/app/analyze/analysis/sampleBin.js [hunk 6]
```diff
@@ -144,39 +144,39 @@ MathFuncts, errGen) {
...
-            // (2b) the array of activity for each node is built by plucking the
-            //      activity `.value` for each sample within this node from the
-            //      `activityCache` [inner .map()]
+            // (2b) the array of activity for each signature is built by
+            //      plucking the activity `.value` for each sample within this
+            //      signature from the `activityCache` [inner .map()]
             function(sampleId) {
-              // FIXME: counting on array order to match node order here
+              // FIXME: counting on array order to match signature order here
               return this.activityCache.get(sampleId)[val.node - 1].value;
```
Comment (dongbohu) [inline interface/src/app/analyze/analysis/sampleBin.js]: I also keep `node` on this line. Not sure whether it is correct.

File: interface/src/app/analyze/analysis/sampleBin.js (hunk 5 above)
Comment (mhuyck) [inline interface/src/app/analyze/analysis/sampleBin.js]: Yes. That is right.

File: interface/src/app/analyze/analysis/sampleBin.js (hunk 6 above)
Comment (mhuyck) [inline interface/src/app/analyze/analysis/sampleBin.js]: Also need to keep it as `node` here, for the same reason as before.

File: interface/src/app/analyze/analysis/analysis.js (hunk 4 above)
Comment (mhuyck) [inline interface/src/app/analyze/analysis/analysis.js]: Hmm. This one might be a bug that's just now showing up because the way your database was loaded is different from mine. If my hunch is correct, you would never have been able to run this clustering the data from your database before.

I'm digging deeper, but might not resolve it before I have to do school pick-up. Will have to clone your branch and try to replicate.

File: interface/src/app/analyze/analysis/sampleBin.js (hunk 6 above)
Comment (mhuyck) [inline interface/src/app/analyze/analysis/sampleBin.js]: It's this line that's biting us. See that `FIXME` above? I figured this would cause a problem some day. Today is the day it gets fixed. ✊

Comment (mhuyck) [inline interface/src/app/analyze/analysis/sampleBin.js]: I think I see what to do. Might be simpler to fix than I feared, but will still take some testing.

Comment (dongbohu) [inline interface/src/app/analyze/analysis/sampleBin.js]: Thanks!

Comment (mhuyck) [inline interface/src/app/analyze/analysis/sampleBin.js]: This problem is actually separate from the renaming you did. If you want to proceed with this PR, let's file a separate issue. I will probably need to change several lines that are scattered around so it would be easier to fix that way anyway.

Comment (mhuyck) [inline interface/src/app/analyze/analysis/sampleBin.js]: I will need help from your database to test my fix when it's ready, though, so don't do anything crazy like rebuild your dev database or anything, okay?

**Comment by mhuyck** [comment]
Okay, I've logged a separate issue for the bug so it doesn't need to hold this up any more. If you haven't found any other issues from this set of changes, I'd say you've completed the UI portion of renaming from "node" to "signature" so this is ready to merge.

**Comment by mhuyck** [comment]
There might be one codeclimate line length issue that you want to fix but I'll leave that up to you. I think we're softening our 80-char line length rule anyway.

**Comment by dongbohu** [comment]
Thanks. I fixed the two lines that are longer than 80 char. Will take another look and if everything is okay, I will merge it.
